// Copyright 2011 Google Inc. All Rights Reserved

// Protobufs for information about a system

syntax = "proto2";

import "grr/proto/jobs.proto";
import "grr/proto/semantic.proto";


// A Process record describing a system process.
message Process {
  option (semantic) = {
    description: "Process information."
  };

  optional uint32 pid = 1;
  optional uint32 ppid = 2;
  optional string name = 3;
  optional string exe = 4;
  repeated string cmdline = 5;
  optional uint64 ctime = 6;
  optional uint32 real_uid = 7;
  optional uint32 effective_uid = 8;
  optional uint32 saved_uid = 9;
  optional uint32 real_gid = 10;
  optional uint32 effective_gid = 11;
  optional uint32 saved_gid = 12;
  optional string username = 13;
  optional string terminal = 14;
  optional string status = 15;
  optional int32 nice = 16;
  optional string cwd = 17;
  optional uint32 num_threads = 18;
  optional float user_cpu_time = 19;
  optional float system_cpu_time = 20;
  optional float cpu_percent = 21;
  optional uint64 RSS_size = 22;
  optional uint64 VMS_size = 23;
  optional float memory_percent = 24;
  repeated string open_files = 25;
  repeated NetworkConnection connections = 26;
}

message NetworkEndpoint {
  optional string ip = 1;
  optional int32 port = 2;
}

message NetworkConnection {
  enum Family {
    INET = 2;
    INET6 = 10;
    INET6_WIN = 23;
    INET6_OSX = 30;
  };
  enum Type {
    UNKNOWN_SOCKET = 0;
    SOCK_STREAM = 1;
    SOCK_DGRAM = 2;
  };

  optional Family family = 1;
  optional Type type = 2 [default = UNKNOWN_SOCKET];
  optional NetworkEndpoint local_address = 3;
  optional NetworkEndpoint remote_address = 4;

  // This enum contains connection states as returned by psutils.
  enum State {
    UNKNOWN = 0;
    CLOSED = 1;
    LISTEN = 2;
    SYN_SENT = 3;
    SYN_RECV = 4;
    ESTABLISHED = 5;
    FIN_WAIT1 = 6;
    FIN_WAIT2 = 7;
    CLOSE_WAIT = 8;
    CLOSING = 9;
    LAST_ACK = 10;
    TIME_WAIT = 11;
    DELETE_TCB = 12;
    // psutils later than v1.0 return "NONE" instead of "UNKNOWN" so we
    // support both.
    NONE = 13;
  };
  optional State state = 5 [default = UNKNOWN];
  optional uint32 pid = 6;
  optional uint64 ctime = 7;
}

// Describe filesystems
message Filesystem {
  optional string device = 1;
  optional string mount_point = 2;
  optional string type = 3;
  optional string label = 4;
};

message MRUFile {
  optional string filename = 1;
  optional uint64 timestamp = 2 [default = 0];
}

message RunKey {
  optional string keyname = 1;
  optional string filepath = 2;
  optional uint64 lastwritten = 3 [default = 0];
}

message WindowsServiceInformation {
  option (semantic) = {
    description: "Describe a windows service or driver. Field names follow"
    " the cybox naming scheme where possible: "
    "http://cybox.mitre.org/language/version2.0.1/xsddocs/objects/Win_Service_Object.html",
  };

  optional string name = 1 [(sem_type) = {
      description: "The name of the service.",
    }];

  optional string description = 2 [(sem_type) = {
      description: "The description string of the service.",
    }];

  optional StatEntry binary = 3 [(sem_type) = {
      description: "The binary on the client. This may also contain an AFF4 "
      "urn to the server copy of the same file.",
    }];

  optional string state = 4 [(sem_type) = {
      description: "Current state of the service.",
    }];

  optional Dict wmi_information = 5 [(sem_type) = {
      description: "Additional information available through WMI.",
    }];

  optional string display_name = 6 [(sem_type) = {
      description: "Displayed name of the service in Windows GUI controls.",
    }];

  optional string driver_package_id = 7 [(sem_type) = {
      description: "Driver Package ID of the service.",
    }];

  enum ErrorControl {
    CRITICAL = 3;
    SEVERE = 2;
    NORMAL = 1;
    IGNORE = 0;
  };

  optional ErrorControl error_control = 8 [(sem_type) = {
      description: "Behaviour of the service on failure to load/initialize",
    }];

  optional string image_path = 9 [(sem_type) = {
      description: "ImagePath of the service.",
    }];

  optional string object_name = 10 [(sem_type) = {
      description: "ObjectName of the service.",
    }];

  enum ServiceMode {
    SERVICE_BOOT_START = 0;
    SERVICE_SYSTEM_START = 1;
    SERVICE_AUTO_START = 2;
    SERVICE_DEMAND_START = 3;
    SERVICE_DISABLED = 4;
  };

  optional ServiceMode startup_type = 11 [(sem_type) = {
      description: "Service start options",
    }];

  enum ServiceType {
    SERVICE_KERNEL_DRIVER = 0x1;
    SERVICE_FILE_SYSTEM_DRIVER = 0x2;
    SERVICE_ADAPTER_ARGS = 0x4;
    SERVICE_WIN32_OWN_PROCESS = 0x10;
    SERVICE_WIN32_SHARE_PROCESS = 0x20;
  };

  optional ServiceType service_type = 12 [(sem_type) = {
      description: "Type of the service.",
    }];

  optional string group_name = 13 [(sem_type) = {
      description: "Specifies the name of the load ordering group of which "
      "this service is a member.",
    }];

  optional string service_dll = 14 [(sem_type) = {
      description: "Name of the DLL instantiated in the service.",
    }];

  optional string registry_key = 15 [(sem_type) = {
      type: "RDFURN",
      description: "The AFF4 location for the service registry key."
    }];

}

message OSXServiceInformation {
  option (semantic) = {
    description: "Describe an OS X service (launchagent/daemon).",
  };

  optional string label = 1 [(sem_type) = {
      description: "Service label.",
    }];
  optional string program = 2 [(sem_type) = {
      description: "Program path run by the service.",
    }];
  repeated string args = 3 [(sem_type) = {
      description: "Program args.  May also include executable if Program is"
                   " not set.",
    }];
  optional uint64 pid = 4 [(sem_type) = {
      description: "Current service PID"
    }];

  optional string sessiontype = 5 [(sem_type) = {
      description: "SessionType: Aqua|StandardIO|Background|LoginWindow"
    }];

  optional uint64 lastexitstatus = 6 [(sem_type) = {
      description: "Last exit return code of service."
    }];

  optional uint64 timeout = 7 [(sem_type) = {
      description: "Service timeout."
    }];

  optional bool ondemand = 8 [(sem_type) = {
      description: "True if the service should start ondemand."
    }];

  repeated string machservice = 9 [(sem_type) = {
      description: "Machservices advertised by this service."
    }];

  repeated string perjobmachservice = 10 [(sem_type) = {
      description: "Machservices advertised by this service."
    }];

  repeated string socket = 11 [(sem_type) = {
      description: "Sockets used by this service."
    }];

  optional string plist = 15 [(sem_type) = {
      type: "RDFURN",
      description: "The AFF4 location for the service plist."
    }];
}

// Describe an installed software package. Cross platform.
message SoftwarePackage {
  optional string name = 1;
  optional string version = 2;
  optional string architecture = 3;
  optional string publisher = 4;
  enum InstallState {
    INSTALLED = 0;
    PENDING = 1;
    UNINSTALLED = 2;
    UNKNOWN = 3;
  };
  optional InstallState install_state = 5 [default = UNKNOWN];
  optional string description = 6;
  optional uint64 installed_on = 7;
  optional string installed_by = 8;

}


// An item from the browser history.
message BrowserHistoryItem {

  optional string url = 1 [(sem_type) = {
      description: "The URL visited.",
    }];

  optional string domain = 2 [(sem_type) = {
      description: "Domain name accessed.",
    }];

  optional uint64 access_time = 3 [(sem_type) = {
      type: "RDFDatetime",
      description: "Time the URL was accessed.",
    }];

  optional string referrer = 4 [(sem_type) = {
      description: "Referrer used for the access",
    }];

  optional string title = 5 [(sem_type) = {
      description: "Title attribute of the page if available.",
    }];

  optional string program_name = 6 [(sem_type) = {
      description: "Name of the program that generated the entry, e.g. Chrome.",
    }];

  optional string source_urn = 7 [(sem_type) = {
      type: "RDFURN",
      description: "Location of object that this entry was found in.",
    }];

  optional string download_path = 8 [(sem_type) = {
      description: "Path of downloaded file.",
    }];
}
